python/pyspark/pandas/series.py

```diff
@@ -5228,9 +5229,20 @@ def asof(self, where: Union[Any, List]) -> Union[Scalar, "Series"]:
             where = [where]
         index_scol = self._internal.index_spark_columns[0]
         index_type = self._internal.spark_type_for(index_scol)
+
+        if np.nan in where:
+            # When `where` is np.nan, pandas returns the last index value.
+            last_index = self._internal.spark_frame.select(F.last(index_scol)).take(1)[0][0]
+            modified_where = [last_index if x is np.nan else x for x in where]
+        else:
+            modified_where = where
+
         cond = [
-            F.max(F.when(index_scol <= SF.lit(index).cast(index_type), self.spark.column))
-            for index in where
+            F.last(
```

Comment:
According to the comment https://github.com/apache/spark/pull/35191/files#r820354270, `F.last` here could also be indeterministic?

Reply:
How about using `F.max_by` with `F.monotonically_increasing_id`?

Reply:
@ueshin  According to https://github.com/apache/spark/pull/35191#discussion_r799222433 . Last is being used , since assumption was dataframe is sorted on index_level_0 .  IF that assumption is not valid , then we cannot use Last . 

cc @itholic

Reply:
The document says `F.last` is non-deterministic.

https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L1337-L1340

Even if the dataframe is sorted beforehand, `F.last` will cause a shuffle because it's an aggregate function, thus the result may be non-deterministic.

Reply:
How about something like https://github.com/ueshin/apache-spark/commit/1b41024b91c9e8dca5d83c5c2793da85f2ba3946?

Reply:
Hi @ueshin . Thx for the suggestion . However, IMHO F.max_by may not work . 

1) Because it will not ignore nulls and also may not work on multiple columns .  For .e.g we have input DF like below 

+------------------------------------------------------------------+-------------------------------------------------------------------+-----------------+
|CASE WHEN (__index_level_0__ <= CAST(5 AS BIGINT)) THEN Koalas END|CASE WHEN (__index_level_0__ <= CAST(20 AS BIGINT)) THEN Koalas END|__natural_order__|
+------------------------------------------------------------------+-------------------------------------------------------------------+-----------------+
|                                                              null|                                                                2.0|                0|
|                                                              null|                                                                1.0|       8589934592|
|                                                              null|                                                               null|      17179869184|
|                                                              null|                                                               null|      25769803776|
+------------------------------------------------------------------+-------------------------------------------------------------------+-----------------+

What we need is the last not null value and if the column have all null then the null value . So the expected output is (which is what F.last is providing) . 


+------------------------------------------------------------------+-------------------------------------------------------------------+-----------------+
|CASE WHEN (__index_level_0__ <= CAST(5 AS BIGINT)) THEN Koalas END|CASE WHEN (__index_level_0__ <= CAST(20 AS BIGINT)) THEN Koalas END|__natural_order__|
+------------------------------------------------------------------+-------------------------------------------------------------------+-----------------+
|                                                              null|                                                                1.0|       8589934592|
+------------------------------------------------------------------+-------------------------------------------------------------------+-----------------+

Please let me know if I have misunderstood . In my earlier commits , I was able to achieve the same with F.explode

Reply:
Could you try https://github.com/ueshin/apache-spark/commit/1b41024b91c9e8dca5d83c5c2793da85f2ba3946?
That is a bit different from what I first suggested, though.
It passes all the tests you added.

Reply:
@ueshin  Thx for the suggestion , have done the changes as suggested